main.cpp

```diff
@@ -125,31 +188,62 @@ void uthread_yield(){
     if (ret_val == 1) return;
 
     // Choose the next thread to execute.
-    int id = current_thread_id;
-    do {
-        id = (id + 1) % num_threads;
-        if(id == current_thread_id){
-            printf("infinite loop!\n");
-        }
-    } while (threads[id].complete);
+    ready_list.push_back(current_thread_id);
+
+    current_thread_id = ready_list.front();
+    ready_list.erase(ready_list.begin());
+
+    siglongjmp(threads[current_thread_id].env,1);
+}
+
+/**
+ * Add calling thread to the waiting list to block. Switch to a
+ * different thread.
+ */
+void thread_switch(){
+
+    int ret_val = sigsetjmp(threads[current_thread_id].env,1);
```

Comment:
This context switch is for threads that called thread_join().  They need to be placed on waiting list.